### ISSUE TITLE
Enhance graph.findConnectedGraph to use type hierarchy

### ISSUE THREAD

**Issue by dselman** [accordproject/concerto-codegen#56]
Currently `graph.findConnectedGraph` does not use the type hierarchy, meaning that types that extend a target type are dropped from a filtered model.

## Context

Using the code below we can "tree shake" the model manager, to remove all types that are not connected to a root/target type, and then render the result as a PlantUML diagram. Although the output is correct, it is aggressive in that if we have a model containing derived types that *could* be used in place of an abstract type the derived types are removed, leaving only the abstract type.

```
if (process.argv.length > 2) {
    const graphVisitor = new Common.ConcertoGraphVisitor();
    const graph = new Common.DirectedGraph();
    mm.accept(graphVisitor, { graph });
    const connectedGraph = graph.findConnectedGraph(process.argv[2]);

    // now create a new model manager that only includes types
    // that are included in the dependency graph
    const filteredModelManager = mm.filter((declaration) =>
      connectedGraph.hasVertex(declaration.getFullyQualifiedName())
    );

    const visitor = new CodeGen.PlantUMLVisitor();
    const parameters = {
      fileWriter: new FileWriter("."),
      showCompositionRelationships: true,
      hideBaseModel: true,
    };
    filteredModelManager.accept(visitor, parameters);
  }
```

## Expected Behavior

It would be very useful to be able to optionally take into account the type hierarchy in the call to `graph.findConnectedGraph` allowing the caller to say "find everything connected to this type, and include all derived types that appear in the model".

## Actual Behavior

Derived types are removed as they are not directly connected in the graph.

## Steps to Reproduce
1. Run the REPL https://replit.com/@dselman/AccordProjectConcerto-Type-Dependencies#index.js
2. Note that the types `person@1.0.0.Dog` and `Cat` are not included, although the abstract type `Pet` which they extend **is**

## Existing issues

- [ ] [Stack Overflow issues](http://stackoverflow.com/tags/concerto)
- [ ] [GitHub Issues](https://github.com/accordproject/concerto/issues)

## Your Environment

* Version used:
* Environment name and version (e.g. Chrome 39, node.js 5.4):
* Operating System and version (desktop or mobile):
* Link to your project:

**Comment by mttrbrts** [comment]
I'd argue that this is a bug. We should be creating edges that follow the type hierarchy in the GraphVisitor.

**Comment by dselman** [comment]
Also Map key and value types.

**Comment by sanketshevkar** [comment]
> I'd argue that this is a bug. We should be creating edges that follow the type hierarchy in the GraphVisitor.

I think I agree with Matt, earlier I thought it could be a addition, but now as I played around it, it feels like a bug.
From the perspective of the mermaid diagram it makes sense and feels like a new addition. But if we want to used the tree shaken model, as given in the example we won't be able instantiate the `abstract Pet` concept and would need ClassDeclarations for `Cat` and `Dog` concepts.

Coming to the point, should we still add the fix as optional? I don't think so. But downside I saw is that after the fix the graph become Directed Cyclic graph, the making the Flowchart representation a bit clunky.

**Comment by sanketshevkar** [comment]
> Similar (but different...) would be decorators that referenced types.

Apart from the bug we identified from today's tech-wg call.
I have used the following [REPL ](https://replit.com/@sanketshevkar/Decorator-TypeReference#index.js) to describe two scenarios.
Referring to both scenarios, won't having a fully qualified type name be more beneficial in decorator arguments object?
It would be easier to get the derived types especially in cases where we might have similar named types in two different namespaces?

**Comment by dselman** [comment]
We should make this consistent with what we are doing for other type references, e.g. TypeIdentifier for super class. What does that look like?

**Comment by sanketshevkar** [comment]
> I also noticed that your code in the REPL seems to be just doing a toString on the object, look at the AST instead, or the API methods.

I am bit confused about the `toString` part

- I am not explicitly doing any `toString`, is there implicit call to `toString` you are referring to?
- Referring to the AST, this how the arguments object looks like this and even here we don't any reference to the fully qualified type.
```
"arguments": [
                  {
                    "$class": "concerto.metamodel@1.0.0.DecoratorTypeReference",
                    "type": {
                      "$class": "concerto.metamodel@1.0.0.TypeIdentifier",
                      "name": "Color"
                    },
                    "isArray": false,
                    "location": {
                      "$class": "concerto.metamodel@1.0.0.Range",
                      "start": {
                        "offset": 117,
                        "line": 9,
                        "column": 17,
                        "$class": "concerto.metamodel@1.0.0.Position"
                      },
                      "end": {
                        "offset": 122,
                        "line": 9,
                        "column": 22,
                        "$class": "concerto.metamodel@1.0.0.Position"
                      }
                    }
                  }
                ]
```

**Comment by sanketshevkar** [comment]
> We should make this consistent with what we are doing for other type references, e.g. TypeIdentifier for super class. What does that look like?

In the AST both superType and decorator TypeIdentifiers have a similar shape.
```
"arguments": [
                  {
                    "$class": "concerto.metamodel@1.0.0.DecoratorTypeReference",
                    "type": {
                      "$class": "concerto.metamodel@1.0.0.TypeIdentifier",
                      "name": "Color"
                    }, .....
```

```
"superType": {
              "$class": "concerto.metamodel@1.0.0.TypeIdentifier",
              "name": "Color"
            }
```
But for super class when I do
`declaration.getSuperClass()` I can get a resolved fully qualified type `workshop@1.0.0.Color`

But for reference type decorators I only get `[{ type: 'Identifier', name: 'Color', array: false }]` when I do `decorator.getArguments()` , which I don't think is incorrect.

But shouldn't the decorator class have some internal methods that can resolve the type reference, similar to the one we have in [classDeclaration ](https://github.com/accordproject/concerto/blob/0d7f108d4961f87782436dd689639ce0f35e97f8/packages/concerto-core/lib/introspect/classdeclaration.js#L155)?

**Comment by mttrbrts** [comment]
> But shouldn't the decorator class have some internal methods that can resolve the type reference, similar to the one we have in [classDeclaration ](https://github.com/accordproject/concerto/blob/0d7f108d4961f87782436dd689639ce0f35e97f8/packages/concerto-core/lib/introspect/classdeclaration.js#L155)?

Yes, you're right, it should. There's a comment in the code that questions this design too!
https://github.com/accordproject/concerto/blob/main/packages/concerto-core/lib/introspect/decorator.js#L80

Can you open an issue for an enhancement to the Introspect API to make it more consistent here, please?
